### PR TITLE
Sdk updates

### PR DESCRIPTION
**Pull request by gaurav1068**
TODO
- [x] update test cases
- [x] check for order signature validation
- [x] add previous function with depreciated warning
- [x] update docs

### CODE REVIEW

File: razorpay/client.py [hunk 1]
```diff
@@ -40,60 +42,96 @@ def __init__(self, session=None, auth=None, **options):
...
         for name, Klass in RESOURCE_CLASSES.items():
             setattr(self, name, Klass(self))
 
+        for name, Klass in UTILITY_CLASSES.items():
+            setattr(self, name, Klass(self))
+
+    def _update_user_agent_header(self, options):
+        user_agent = "{}{}".format('Razorpay-Python/',  self._get_version())
+
+        if 'headers' in options:
```
Comment (venkatvghub) [inline razorpay/client.py]: What about headers already having useragent?

File: razorpay/client.py [hunk 2]
```diff
@@ -40,60 +42,96 @@ def __init__(self, session=None, auth=None, **options):
...
         Dispatches a request to the Razorpay HTTP API
         """
-        options = self._merge_options(options)
-        url = "{}{}".format(options['base_url'],  path)
-        request_options = self._parse_request_options(options)
+        base_url = self.DEFAULTS['base_url']
+
+        if 'base_url' in options:
+            base_url = options['base_url']
+            del(options['base_url'])
```
Comment (venkatvghub) [inline razorpay/client.py]: why is this needed here?

File: razorpay/client.py [hunk 3]
```diff
@@ -40,60 +42,96 @@ def __init__(self, session=None, auth=None, **options):
...
+
+        options = self._update_user_agent_header(options)
+
+        url = "{}{}".format(base_url,  path)
         response = getattr(self.session, method)(url, auth=self.auth,
                                                  verify=self.cert_path,
-                                                 **request_options)
-        if response.status_code == 200:
+                                                 **options)
+        if ((response.status_code >= 200) and (response.status_code < 300)):
```
Comment (venkatvghub) [inline razorpay/client.py]: please use a constant for 200 and 300

File: razorpay/client.py [hunk 4]
```diff
@@ -40,60 +42,96 @@ def __init__(self, session=None, auth=None, **options):
...
+            code = ""
             json_response = response.json()
             if 'error' in json_response:
                 if 'description' in json_response['error']:
                     msg = json_response['error']['description']
-            if response.status_code == 400:
+                if 'code' in json_response['error']:
+                    code = json_response['error']['code']
+
+            if str.upper(code) == 'BAD_REQUEST_ERROR':
```
Comment (venkatvghub) [inline razorpay/client.py]: use constants please.

File: razorpay/client.py [hunk 5]
```diff
@@ -40,60 +42,96 @@ def __init__(self, session=None, auth=None, **options):
...
                 raise BadRequestError(msg)
-            if response.status_code == 401:
-                raise NoAuthorizationError(msg)
-            if response.status_code == 404:
-                raise NotFoundError(msg)
-            elif response.status_code >= 500 and response.status_code < 600:
+            elif str.upper(code) == 'GATEWAY_ERROR':
```
Comment (venkatvghub) [inline razorpay/client.py]: same here. Use constants please.

File: razorpay/client.py [hunk 6]
```diff
@@ -102,66 +140,6 @@ def _update_request(self, data, options):
         if 'headers' in options:
             options['headers']['Content-type'] = 'application/json'
         else:
-            options['headers'] = {'Content-type' : 'application/json'}
+            options['headers'] = {'Content-type': 'application/json'}
```
Comment (venkatvghub) [inline razorpay/client.py]: What if the content type is already set again?

File: razorpay/resources/card.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
+from .base import Resource
+from .Url import URL
+
+
+class Card(Resource):
+    def __init__(self, client=None):
```
Comment (venkatvghub) [inline razorpay/resources/card.py]: add spacing between assignments all through please.

File: razorpay/resources/payment.py [hunk 8]
```diff
@@ -42,10 +46,10 @@ def capture(self, payment_id, amount, **kwargs):
             Payment dict after getting captured
         """
         url = "{}/{}/capture".format(self.base_url, payment_id)
-        data = {'amount': amount}
+        data['amount'] = amount
```
Comment (venkatvghub) [inline razorpay/resources/payment.py]: what was the issue with `data = {'amount': amount}`?

File: razorpay/resources/order.py [hunk 9]
```diff
@@ -27,10 +32,15 @@ def fetch(self, order_id, **kwargs):
...
+        return super(Order, self).fetch(order_id, data, **kwargs)
+
+    def fetch_all_payments(self, order_id, data={}, **kwargs):
+        warnings.warn("Will be Deprecated in next release, use payments",
+                      DeprecationWarning)
+        self.payments(order_id, data, **kwargs)
+
 
-    def fetch_all_payments(self, order_id, **kwargs):
+    def payments(self, order_id, data={}, **kwargs):
```
Comment (venkatvghub) [inline razorpay/resources/order.py]: reduce extra linespace before `payments(..)`

File: razorpay/client.py (hunk 1 above)
Comment (gaurav1068) [inline razorpay/client.py]: they should be overridden, this is by design.

File: razorpay/client.py (hunk 2 above)
Comment (gaurav1068) [inline razorpay/client.py]: its throwing an exception, if it is not deleted

File: razorpay/resources/order.py [hunk 10]
```diff
@@ -53,7 +63,7 @@ def create(self, data={}, **kwargs):
                 'currency' : Currency used in Order
                 'receipt' : Receipt Id for the order
                 'notes' : key value pair as notes
-                'payment_capture'  : 0/1 if payment should be auto captured or not
+                'payment_capture': 0/1 if payment should be auto captured or not
```
Comment (venkatvghub) [inline razorpay/resources/order.py]: pep8: line too long. Please reduce or split into multiple lines.

File: razorpay/utility/utility.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+        payment_id = parameters['payment_id']
+        razorpay_signature = parameters['razorpay_signature']
+        msg = "{}|{}".format(order_id, payment_id)
+        dig = hmac.new(key=self.client.auth[1],
+                       msg=msg,
+                       digestmod=hashlib.sha256)
+
+        generated_signature = dig.hexdigest()
+
+        if hmac.compare_digest(generated_signature, razorpay_signature) == False:
```
Comment (venkatvghub) [inline razorpay/utility/utility.py]: use `if cond is False:` or `if not cond:`

Comment (venkatvghub) [inline razorpay/utility/utility.py]: pep8: this line is too long. Split into multiple lines please.

File: tests/test_client_token.py [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+        super(TestClientCustomer, self).setUp()
+        self.base_url = '{}/customers'.format(self.base_url)
+
+    @responses.activate
+    def test_tokens_all(self):
+        result = mock_file('token_collection')
+        url = '{}/{}/tokens'.format(self.base_url, self.customer_id)
+        responses.add(responses.GET, url, status=200, body=json.dumps(result),
+                      match_querystring=True)
+        self.assertEqual(self.client.token.all(self.customer_id, self.token_id),
```
Comment (venkatvghub) [inline tests/test_client_token.py]: pep8: line too long. Please split into multiple lines

File: razorpay/client.py (hunk 6 above)
Comment (gaurav1068) [inline razorpay/client.py]: then it will fail, as we are sending dictionary object in invoice and without this header our req will fail

File: razorpay/resources/payment.py (hunk 8 above)
Comment (gaurav1068) [inline razorpay/resources/payment.py]: it will override data

File: razorpay/resources/card.py (hunk 7 above)
Comment (gaurav1068) [inline razorpay/resources/card.py]: py lint gives Error,

**Comment by gaurav1068** [comment]
screen shot 2017-02-15 at 10 17 34 pm
Will write create test cases

**Comment by gaurav1068** [comment]
Run coverage using nosetests --with-coverage --cover-package=razorpay
screen shot 2017-02-16 at 1 15 39 pm

**Comment by captn3m0** [comment]
@gaurav1068 I can setup coveralls here if you want in a different PR

**Comment by gaurav1068** [comment]
final coverage percentage 97%

File: razorpay/constants.py [hunk 13]
```diff
@@ -0,0 +1,15 @@
+class HTTP_STATUS_CODE(object):
+    OK = 200
+    REDIRECT = 300
+    ORDER_URL = "/orders"
```
Comment (venkatvghub) [inline razorpay/constants.py]: Can you keep the status codes in one class and the urls in another

**Comment by venkatvghub** [comment]
@gaurav1068 I haven't looked into the functionality and assuming it looks good. Do you want me to look into it?
